### DISCUSSION TITLE
Have a problem with navigation in IOS App

### DISCUSSION THREAD

**Discussion by Gra-Dus** [arkivanov/Decompose#616] [Q&A] [answered]
Hello. I develop kmm module. Android App works good with navigation. But in IOS App navigation sending warning "A navigationDestination for “UIOSChild” was declared earlier on the stack. Only the destination declared closest to the root view of the stack will be used." I dont undestand how to solve that.

**Comment by arkivanov** [comment]
It's hard to say without a reproducer. But it sounds like you have duplicated distinctions in your stack, or something like that.

**Reply by Gra-Dus** [reply]
it looks like somehow my childs defines like "UIOSChild", so comparsion give me that warning. I have few nested stack navigation. That warning firstly comes when i replace splash to Auth. After that when i try navigate to other stack navigation, I comes to initial component and go back to "Auth".

**Reply by arkivanov** [reply]
What is `UIOSChild`? Also how do you manage the navigation on iOS?

**Reply by Gra-Dus** [reply]
Based on the search in the framework code, I can say that this is a Child.
![Screenshot 2024-01-24 at 3 19 37 PM](https://github.com/arkivanov/Decompose/assets/65005478/ea341009-eb70-40d6-be50-547daecaa77a)
For navigation we use StackView from sample. I think somewhere in NavigationStack happens comparsion.

**Reply by arkivanov** [reply]
Ok, it seems corresponds to [Child](https://github.com/arkivanov/Decompose/blob/master/decompose/src/commonMain/kotlin/com/arkivanov/decompose/Child.kt) sealed class from Decompose. With the reproducer code it's hard to say what's the cause. But I think it's related to nested navigation (`navigationDestination(for: Child<AnyObject, T>.self)` being repeated). Would it be possible for you to create a runnable reproducer?

**Reply by Gra-Dus** [reply]
https://github.com/Gra-Dus/UIOSChildSampleError

**Reply by arkivanov** [reply, marked as answer]
Thanks for the reproducer! I don't have those warnings, but the issue is still reproducible. When I click on "goToReg", the new screen is pushed and then immediately popped back. Is this the issue you are experiencing?

If so, for now I don't have any solution, this looks like a bug in `NavigationStack` when it's nested. One thing that I noticed is that [StackInteropView](https://github.com/arkivanov/Decompose/blob/d1036691812edb6b1dba166f68fc96bf76ec08d0/sample/app-ios/app-ios/DecomposeHelpers/StackView.swift#L32) works fine. So maybe as a workaround use only that one.

**Reply by arkivanov** [reply]
In https://forums.developer.apple.com/forums/thread/665369?answerId=719452022#719452022 the following is mentioned:

> Definitely broken in iOS16. It appears (from my reading) that this is the way things work. Any change to an ObservedObject outside of a NavigationStack or sheet will cause the parent view to be rebuilt, which in turn wipes the stack or sheet.

This sounds like exactly what you are experiencing.

**Reply by arkivanov** [reply]
Also, this [post](https://forums.developer.apple.com/forums/thread/737671?answerId=765059022#765059022) says:

> The intended way to use it is to have one top-level NavigationStack

Which is sad if it's true.

**Reply by Gra-Dus** [reply]
Thank you for help.

**Reply by arkivanov** [reply]
You can try this variant of StackView by @AJIEKCX: https://gist.github.com/AJIEKCX/e1c50fb3b1c0496138bcdbb2779e3898

**Comment by FelixFalkovsky** [comment]
No solution to the problem ?

**Reply by arkivanov** [reply]
There were a few options discussed above. Worth trying some of them

**Reply by FelixFalkovsky** [reply]
Спасибо последнее решение помогло !
